### PR TITLE
armv8-m: making full SRAM run-time partitioning (MPU gap filling) optional

### LINKED ISSUE #19067: non-overlapping MPU gap-filling needs to be optional
ARMv8, and ARC MPU v3 are newer MPUs which allow for great flexibility on MPU region sizes, but regions are not allowed to overlap.

Currently, MPU regions are set up to control supervisor mode access, so that supervisor mode reads/writes to invalid memory generate faults. Implementing overlapping regions for memory domain requires that these base regions be split.

Unfortunately, this gap-filling has drawbacks:
1. The number of MPU regions available to applications is cut roughly in half, due to the need for additional regions to fill in the gaps
2. The gap-filling introduces overhead, currently during context switch

The overhead can be looked at (see #15223) but halving the number of free regions may in a lot of cases be a dealbreaker for applications.

This gap-filling is NOT needed to properly support user mode. This is a feature to catch stray memory access in supervisor mode and is orthogonal to CONFIG_USERSPACE requirements.

Add a Kconfig option for these MPUs which allows this to be disabled, maximizing the number of memory domain partitions available to applications.

### PR DESCRIPTION
**Pull request by ioannisg**
Addresses #19067 for ARM.

It should be working atm.
I would like a review on the Kconfig symbol name I've used, and the default behavior.
I am also going to add a test of the feature,

### CODE REVIEW

File: arch/arm/core/cortex_m/mpu/Kconfig [hunk 1]
```diff
@@ -62,6 +62,30 @@ config MPU_ALLOW_FLASH_WRITE
 	help
 	  Enable this to allow MPU RWX access to flash memory
 
+config MPU_SKIP_BACKGROUND_SRAM_FILLING
```
Comment (andrewboie) [inline arch/arm/core/cortex_m/mpu/Kconfig]: can we change the semantics of this Kconfig?
it would be more natural to have a "config MPU_GAP_FILLING" that is default N. And probably placed not in ARM space, since ARC also does this.

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: I see the point of pushing this up in the arch KConfig; I'll do that.

Regarding the semantics; I am ok with changing the semantics; not so fond of the name. How about MPU_FULL_BACKGROUND_SRAM_PARTITIONING? (and default to N)

@andrewboie, it is important to agree on the default behavior. So you think that the default behavior on ARMv8-M (and ARC v2) should be to _not_ enforce this gap filling? Or the other way around?

**Comment by andrewboie** [comment]
I think it should be off by default -- an extra feature that can be enabled if users want it. It's not required for user mode at any rate.

As far as the naming of the Kconfig, MPU_FULL_BACKGROUND_SRAM_PARTITIONING is rather long winded, but sure, no complaints if you want to call it that.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by ioannisg** [comment]
I want to have this in for Zephyr v2.1 release; is it possible to have additional reviewing (besides @andrewboie ) on the introduced Kconfig option at least?

**Review by wentongwu — APPROVED** [review]

**Comment by ulfalizer** [comment]
Note that the direct (`depends on`) dependencies of symbols that are defined in multiple locations get ORed together, so there's no way to add extra dependencies by defining a symbol more than once.

The dependencies of `MPU_SRAM_GAP_FILLING` end up as this (copied frrom the menuconfig):

```
Direct dependencies (=n):
     USERSPACE(=n) && ARM_MPU(=n) && CPU_HAS_MPU(=y) && ARM(=y)  (=n)
  || MPU_REQUIRES_NON_OVERLAPPING_REGIONS(=n)
```

If `MPU_REQUIRES_NON_OVERLAPPING_REGIONS` is `y`, it will always be possible to enable `MPU_SRAM_GAP_FILLING`. Is that okay here?

A tip is to go into the menuconfig and jumping to each of the definition locations of the symbol. That might help with figuring out how it works. Basically, the definition in `arch/Kconfig` will always be visible as long as `MPU_REQUIRES_NON_OVERLAPPING_REGIONS` is `y`, allowing the symbol to be toggled there.

**Comment by ioannisg** [comment]
> Note that the direct (`depends on`) dependencies of symbols that are defined in multiple locations get ORed together, so there's no way to add extra dependencies by defining a symbol more than once.
> 
> The dependencies of `MPU_SRAM_GAP_FILLING` end up as this (copied frrom the menuconfig):
> 
> ```
> Direct dependencies (=n):
>      USERSPACE(=n) && ARM_MPU(=n) && CPU_HAS_MPU(=y) && ARM(=y)  (=n)
>   || MPU_REQUIRES_NON_OVERLAPPING_REGIONS(=n)
> ```
> 
> If `MPU_REQUIRES_NON_OVERLAPPING_REGIONS` is `y`, it will always be possible to enable `MPU_SRAM_GAP_FILLING`. Is that okay here?
> 
> A tip is to go into the menuconfig and jumping to each of the definition locations of the symbol. That might help with figuring out how it works. Basically, the definition in `arch/Kconfig` will always be visible as long as `MPU_REQUIRES_NON_OVERLAPPING_REGIONS` is `y`, allowing the symbol to be toggled there.

Thanks so much @ulfalizer for this feedback. 

>> If `MPU_REQUIRES_NON_OVERLAPPING_REGIONS` is `y`, it will always be possible to enable `MPU_SRAM_GAP_FILLING`. Is that okay here?

Not exactly. I need `MPU_SRAM_GAP_FILLING `to have a direct dependency on `MPU_REQUIRES_NON_OVERLAPPING_REGIONS`, but I want it to be defined only if `USERSPACE `is also defined. I had the impression that dependencies are ANDed instead of ORed. So what's your proposal?

File: arch/arm/core/cortex_m/mpu/Kconfig [hunk 2]
```diff
@@ -62,6 +62,31 @@ config MPU_ALLOW_FLASH_WRITE
 	help
 	  Enable this to allow MPU RWX access to flash memory
 
+config MPU_SRAM_GAP_FILLING
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/mpu/Kconfig]: @ulfalizer can we put the definition into `if USERSPACE`  block, to make it defined only if USERSPACE is also defined? Then, what will happen with the dependency on `MPU_REQUIRES_NON_OVERLAPPING_REGIONS`?

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: (Note that `if FOO` is just a shorthand for adding `depends on FOO` to the symbol(s) within it. `if` and `depends on` are exactly the same thing. I think this is something that confuses people re. Kconfig, where people assume `if` does more than that. :)

If a symbol is defined in multiple locations and should depend on `USERSPACE`, then all definition locations need to have `depends on USERSPACE` (or be within an `if USERSPACE` -- same thing), since the dependencies get ORed together.

Personally, I'd try to refactor it so that there's just a single definition in `arch/Kconfig`, maybe along with some helper symbols.

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: >Personally, I'd try to refactor it so that there's just a single definition in arch/Kconfig, maybe along with some helper symbols.

That's hard; I raised the definition up in arch/Kconfig, as it is essentially a cross-arch option, but I want to keep re-defining it in arch/arm/cortex-m/mpu/Kconfig , so I can have the additional help text that is ARMv8-M specific. But it is becoming clear from your comments that dependencies are always ORed so I cannot add additional dependencies here (so I need to move depends on USERSPACE up in the top-level definition, correct?)

So how can I do it in a way it satisfies you @ulfalizer ? :)

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: These two are 100% identical:

```
if FOO
if BAR

config FOO
	bool "foo"
	depends on BAZ

endif
endif
```

```
config FOO
	bool "foo"
	depends on FOO && BAR && BAZ
```

`if` does not "cancel" the definition, but only modified the dependencies, like `depends on` does. When a symbol is defined more than once, the dependencies get `||`'d together.

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: A tip is to play around with this in `menuconfig` btw, because it'll make it clear how multiple definitions work:

```
config DEP1
	bool "dep1"

config DEP2
	bool "dep2"

config MULTIDEF
	bool "first multidef prompt"
	depends on DEP1

config MULTIDEF
	bool "second multidef prompt"
	depends on DEP2
```

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: > These two are 100% identical:
> 
> ```
> if FOO
> if BAR
> 
> config FOO
> 	bool "foo"
> 	depends on BAZ
> 
> endif
> endif
> ```
> 
> ```
> config FOO
> 	bool "foo"
> 	depends on FOO && BAR && BAZ
> ```
> 
> `if` does not "cancel" the definition, but only modified the dependencies, like `depends on` does. When a symbol is defined more than once, the dependencies get `||`'d together.

Understood. So can you help me finding a solution that satisfies my needs?
- I want MPU_SRAM_GAP_FILLING to be defined high in arch/Kconfig, as it may be used by multiple ARCHes
- I want it to depend on MPU_REQUIRES_NON_OVERLAPPING_REGIONS **and** USERSPACE
- But: I want the help text to be somewhat ARM-specific, that's why I want to have the option re-defined in the ./arm sub-directory. 

So you suggest keeping the solution **as is** and simply add the same list of dependencies in the two definitions of the symbol?

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: > but I want to keep re-defining it in arch/arm/cortex-m/mpu/Kconfig, so I can have the additional help text that is ARMv8-M specific

I'd have a single definition and some ARM-specific notes in it (well sectioned off). Think it's less of a mess than dealing with multiple definition locations.

Otherwise, the only option is to add `depends on USERSPACE` to all definitions, if the symbol should (always) depend on `USERSPACE`.

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: Another option would be to have multiple definitions, but only putting them in `arch`-specific Kconfig files, and not in the generic `arch/Kconfig`.

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: > So you suggest keeping the solution **as is** and simply add the same list of dependencies in the two definitions of the symbol?

Yep, would have to do that, if the definitions are kept as-is.

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: @ulfalizer thanks. There's also an option to have MPU_SRAM_GAP_FILLING defined in arch/Kconfig,
```
config MPU_SRAM_GAP_FILLING
    bool "bla bla bla, to make it visible "
    depends on MPU_REQUIRES_NON_OVERLAPPING_REGIONS
    help
       Some generic cross-ARCH text
```

 and have an arm-specific option in arm/cortex-m/mpu/Kconfig as:
```
config ARM_MPU_SRAM_GAP_FILLING
    bool
    depends on USERSPACE
    depends on MPU_SRAM_GAP_FILLING
    default y
    help 
      Some ARM-specific text
```

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: How about the last option??

>Another option would be to have multiple definitions, but only putting them in arch-specific Kconfig files, and not in the generic arch/Kconfig.

We thought of not doing this, as it is essentially the same functionality used in different ARCHes

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: Yeah, that might work too. One drawback is that it'll still be possible to enable `MPU_SRAM_GAP_FILLING` when `USERSPACE` is `n`. Would that be confusing?

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: No. If someone does it, it will simply be ignored by the underlying, ARM-specific option, which adds the USERSPACE dependency, and that's totally OK.

Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: The point is that : only the ARCH-specific option will be used in the code. the top-arch option will only be used in Kconfig, to enable the ARM-specific option.

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: Was thinking from an interface design perspective. There'll be a button you toggle that says "MPU SRAM gap filling" or the like, but it won't do anything unless you also toggle "User mode threads".

Feels like it gets lost sometimes that Kconfig defines a configuration interface. :)

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: As long as that wouldn't be weird to users, it'll work at least.

Comment (ulfalizer) [reply arch/arm/core/cortex_m/mpu/Kconfig]: My preferred solution would be either

1) A single definition in `arch/Kconfig`, with some arch-specific notes in the help text, and maybe some helper symbols to adjust it for different arch capabilities

2) Have multiple arch-specific definitions, and skip the definition in `arch/Kconfig`

I won't impose anything though. Just keep the multiple-definition dependency gotchas in mind.

File: arch/arm/core/cortex_m/mpu/Kconfig [hunk 3]
```diff
@@ -62,6 +62,29 @@ config MPU_ALLOW_FLASH_WRITE
...
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: >Why do we need this?
All the gap filling does is restrict access to supervisor mode threads, which we don't need for user mode.
I would rather that this option, when not enabled, does no automatic gap filling at all.

We do not need this, @andrewboie, _right now_ because any static regions we enforce at boot time:
- code (mandatory)
- execute-from-sram (optional)
- no-cache (optional)
- read-only (optional)
- gcov (optional)
- ....
are properly partitioning the memory map, i.e. they do not leave gaps in between them. So, not allowing gaps does not make us waste regions. Therefore, the MPU driver behavior when programming the static regions is the same, regardless of whether MPU_SRAM_GAP_FILLING is set or unset.

In other words, we only need (_and should allow_) to have gaps in the sram part where we have dynamic regions defined (i.e. the kernel sram area). So I wanted to restrict the Kconfig option, to reflect _exactly_ this. I don't see why we need the Kconfig option to mean something more generic; I have no plans to allow the MPU driver to skip gap filling outside the kernel SRAM, anyways.

Do you anticipate the need (in the near future) of a Zephyr memory map organization (done at linker level), which will leave gaps between the static sections?

[the remark this answers is not in the document]
Comment (andrewboie) [reply arch/arm/core/cortex_m/mpu/Kconfig]: > And what do you propose for the Kconfig option ?

CONFIG_MPU_GAP_FILLING?

[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: @andrewboie that's fine with me, obviously
But do you approve the help text below the option? Also copying it here for reference:

```
          This Kconfig option instructs the MPU driver to enforce
	  a full kernel SRAM partitioning, when it programs the
	  dynamic MPU regions (user thread stack, PRIV stack guard
	  and application memory domains) during context-switch. We
	  allow this to be a configurable option, in order to be able
	  to switch the option off and have an increased number of MPU
	  regions available for application memory domain programming.
	  Note that an increased number of MPU regions should only be
	  required when building with USERSPACE support.

```

[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: > CONFIG_MPU_GAP_FILLING

Fine with me (just have to remove SRAM from the name)

[the remark this answers is not in the document]
Comment (ioannisg) [reply arch/arm/core/cortex_m/mpu/Kconfig]: Done now @andrewboie

**Comment by ulfalizer** [comment]
> @ulfalizer could you come back to this, briefy?
> FYI, I integrated your suggestion of _not_duplicating the definition of the Kconfig option - it is really cleaner. So, I introduced an internal hidden ARM-only option that takes the configuration enforced by the user. The internal option is used in source code.

Personally, I think I'd either have a single shared symbol in `arch/Kconfig` with some platform-specific notes in the help text, or separate symbols per `arch/<platform>/mpu/Kconfig` (maybe along with factoring out common ones into some `Kconfig.*_mpu` files or the like).

The main motivation is to avoid complicated multiple definition stuff where not many people know how it works (and that's easy to get wrong re. dependencies).

I won't block anything here though, so go with whatever you prefer. :)

Please also check how things look in the `menuconfig`. With the current approach, I suspect the help text for `ARM_MPU_SRAM_GAP_FILLING` might be hard to notice, because it's not what you get when you press `?` on the configurable symbol in the `menuconfig`.

**Review by aescolar — DISMISSED** [review]
@ioannisg note that you still have in this branch the sanitycheck hotfix which is already in master (I do not understand why github is not complaining about the conflict and let's one press the button..) Please disregard this review on your own when you fix it (no need to ping me back)

**Comment by ioannisg** [comment]
> @ioannisg note that you still have in this branch the sanitycheck hotfix which is already in master (I do not understand why github is not complaining about the conflict and let's one press the button..) Please disregard this review on your own when you fix it (no need to ping me back)

The commit is present because I did not rebase.
There is no conflict because the commit in this PR is identical to what was merged (If I understand GH properly)

**Comment by ioannisg** [comment]
> > @ulfalizer could you come back to this, briefy?
> > FYI, I integrated your suggestion of _not_duplicating the definition of the Kconfig option - it is really cleaner. So, I introduced an internal hidden ARM-only option that takes the configuration enforced by the user. The internal option is used in source code.
> 
> Personally, I think I'd either have a single shared symbol in `arch/Kconfig` with some platform-specific notes in the help text, or separate symbols per `arch/<platform>/mpu/Kconfig` (maybe along with factoring out common ones into some `Kconfig.*_mpu` files or the like).
> 
> The main motivation is to avoid complicated multiple definition stuff where not many people know how it works (and that's easy to get wrong re. dependencies).
> 
> I won't block anything here though, so go with whatever you prefer. :)
> 
> Please also check how things look in the `menuconfig`. With the current approach, I suspect the help text for `ARM_MPU_SRAM_GAP_FILLING` might be hard to notice, because it's not what you get when you press `?` on the configurable symbol in the `menuconfig`.

@ulfalizer I finally went along with your preference here, i.e. removed the internal option, and added the ARMv8-M specific explanatory text in the `config ARM_MPU` option.

**Review by tyrantlizard77 — APPROVED** [review]
